datalad/interface/rerun.py

```diff
@@ -0,0 +1,277 @@
+            default="HEAD",
+            constraints=EnsureStr()),
+        dataset=Parameter(
+            args=("-d", "--dataset"),
+            doc="""specify the dataset from which to rerun a recorded command.
+            If no dataset is given, an attempt is made to identify the dataset
+            based on the current working directory. If a dataset is given,
+            the command will be executed in the root directory of this
+            dataset.""",
+            constraints=EnsureDataset() | EnsureNone()),
+        message=save_message_opt,
+        branch=Parameter(
+            metavar="NAME",
+            args=("-b", "--branch",),
+            doc="create and checkout this branch before rerunning the commands.",
+            constraints=EnsureStr() | EnsureNone()),
+        onto=Parameter(
+            metavar="base",
+            args=("--onto",),
+            doc="""start point for rerunning the commands.  If not specified, commands
+            are executed at HEAD.  This option can be used to specify
+            an alternative start point, which will be checked out with
+            the branch name specified with --branch or in a detached
+            state otherwise.  As a special case, an empty value for
+            this option means to use the parent of the first commit
+            (e.g., 'a' for the range 'a..b' or 'a^' for the revision
+            'a') as the starting point.""",
+            constraints=EnsureStr() | EnsureNone()),
+        root=Parameter(
+            args=("--root",),
+            action="store_true",
+            doc="""rerun commands from all commits reachable from revision rather than
+            running only the command from the revision.  This flag is
+            incompatible with using a range for the revision the
+            argument. In other words, run all the commands that would
+            be shown by `git log <revision>`."""),
+        # TODO
+        # --list-commands
+        #   go through the history and report any recorded command. this info
+        #   could be used to unlock the associated output files for a rerun
+    )
+
+    @staticmethod
+    @datasetmethod(name='rerun')
+    @eval_results
+    def __call__(
+            revision="HEAD",
+            dataset=None,
+            branch=None,
+            message=None,
+            onto=None,
+            root=False):
+
+        ds = require_dataset(
+            dataset, check_installed=True,
+            purpose='rerunning a command')
+
+        lgr.debug('rerunning command output underneath %s', ds)
+
+        from datalad.tests.utils import ok_clean_git
+        try:
+            ok_clean_git(ds.path)
+        except AssertionError:
+            yield get_status_dict(
+                'run',
+                ds=ds,
+                status='impossible',
+                message=('unsaved modifications present, '
+                         'cannot detect changes by command'))
+            return
```

Comment:
We would not even expect changes. But some logic further down could change branch etc -- that would be the problem.